crates/storage/src/pull_or_init.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright 2018-2022 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The module helps during compilation time decide which pull mechanism to use.
+//! If the type implements [`OnCallInitializer`](crate::traits::OnCallInitializer) trait,
+//! it will use `pull_storage` with combination `OnCallInitializer::initialize`(if the pull failed).
+//! Otherwise, it will use only `pull_storage` as a default behavior.
+//!
+//! [`OnCallInitializer`](crate::traits::OnCallInitializer) allows initialize the
+//! type on demand. For more information, check the documentation of the trait.
+
+use crate::traits::{
+    pull_storage,
+    OnCallInitializer,
+};
+use ink_primitives::{
+    traits::Storable,
+    Key,
+};
+
+/// Part of Autoref-Based Specialization. It is a wrapper around the type to support autoref
+/// specialization.
+pub struct PullOrInit<T: Storable> {
+    marker: core::marker::PhantomData<fn() -> T>,
+}
+
+impl<T: OnCallInitializer + Storable> PullOrInit<T> {
+    #[allow(dead_code)]
+    pub fn pull_or_init(key: &Key) -> T {
+        let maybe_instance = ink_env::get_contract_storage::<Key, T>(key);
+        match maybe_instance {
+            Ok(None) | Err(_) => {
```

Comment:
Possible attack vector: this could be triggered by a call to an unrestricted `message`? I'm not sure what the attack would be but it means that any user could have control about when this is called if it has not already been initialized.

So not as safe as a constructor.

Reply:
Maybe could add a check to ensure only the owner can cause this to be invoked

Reply:
How do we know who is the owner of the contract?=)

Reply:
Is it a possible attack vector though?

Reply:
It can be called on any message(read and write via transaction) one time per contract(while storage is empty). If the attacker will somehow corrupt the storage, then yes, he again can call the initialization.

Reply:
> one time per contract(while storage is empty)

Also possibly after a contract is upgraded and the storage layout has changed?

Maybe it's not that bad, it just feels like a potential footgun to have a hook which allows any user to run the migration code, even unintentionally.

Reply:
Yes, also after the upgrade=) The user who starts using it should know how to use it. We plan to highlight in the OpenBrush documentation about upgradable contracts

Reply:
Okay, I'd be curious to know what the other guys think about this @cmichi @HCastano @athei. I'm not an expert in upgradable contract patterns.

> The user who starts using it should know how to use it.

Yes of course, which is why we should think very carefully when adding such features.

Reply:
@xgreenx we should pull this out of this PR and discuss it elsewhere. It deserves a bit more thinking. We can get rid of the proxy examples for now

Reply:
Indeed after reading up properly about the diamond patterns etc. it will be really useful to have a feature like this. Agree it is worth discussing as a feature on its own.